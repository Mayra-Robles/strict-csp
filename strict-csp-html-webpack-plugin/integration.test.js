const webpack = require('webpack');
const path = require('path');
const fs = require('fs');
const cheerio = require('cheerio');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const StrictCspHtmlWebpackPlugin = require(path.resolve(__dirname, './plugin.js'));

const runWebpack = (options, done) => {
  const config = {
    context: __dirname,
    mode: 'production',
    entry: {
      library1: './test-fixture/src/library1.js',
      app: './test-fixture/src/app.js',
      library2: './test-fixture/src/library2.js',
    },
    output: {
      path: path.resolve(__dirname, 'test-fixture/dist'),
      filename: '[name].bundle.js',
    },
    plugins: [
      new HtmlWebpackPlugin({
        template: './test-fixture/src/template.html',
      }),
      new StrictCspHtmlWebpackPlugin(HtmlWebpackPlugin, options),
    ],
  };

  webpack(config, (err, stats) => {
    if (err) {
      return done(err);
    }
    if (stats.hasErrors()) {
      return done(new Error(stats.toJson().errors.map((e) => e.message).join('\n')));
    }

    const outputFile = path.resolve(
      __dirname,
      'test-fixture/dist/index.html'
    );
    const outputHtml = fs.readFileSync(outputFile, 'utf8');
    done(null, outputHtml);
  });
};

describe('StrictCspHtmlWebpackPlugin Integration Test', () => {
  it('should build successfully without Trusted Types', (done) => {
    runWebpack({}, (err, outputHtml) => {
      if (err) return done(err);
      const $ = cheerio.load(outputHtml);
      const metaTag = $('meta[http-equiv="Content-Security-Policy"]');
      expect(metaTag.length).toBe(1);
      const cspContent = metaTag.attr('content');
      expect(cspContent).not.toContain('require-trusted-types-for');
      expect(outputHtml).toMatchSnapshot();
      done();
    });
  });

  it('should build successfully with Trusted Types enabled', (done) => {
    runWebpack({ trustedTypes: true }, (err, outputHtml) => {
      if (err) return done(err);
      const $ = cheerio.load(outputHtml);
      const metaTag = $('meta[http-equiv="Content-Security-Policy"]');
      expect(metaTag.length).toBe(1);
      const cspContent = metaTag.attr('content');
      expect(cspContent).toContain("require-trusted-types-for 'script'");
      expect(outputHtml).toMatchSnapshot();
      done();
    });
  });

  it('should build successfully with Trusted Types in report-only mode', (done) => {
    runWebpack(
      {
        trustedTypes: 'report-only',
        reportUri: 'https://example.com/report',
      },
      (err, outputHtml) => {
        if (err) return done(err);
        const $ = cheerio.load(outputHtml);
        const metaTag = $('meta[http-equiv="Content-Security-Policy"]');
        expect(metaTag.length).toBe(1);
        const cspContent = metaTag.attr('content');
        expect(cspContent).toContain("require-trusted-types-for 'script'");
        const loaderScript = $('script:not([src])').html();
        expect(loaderScript).toContain('const generateAndSendReport');
        expect(outputHtml).toMatchSnapshot();
        done();
      }
    );
  });
});