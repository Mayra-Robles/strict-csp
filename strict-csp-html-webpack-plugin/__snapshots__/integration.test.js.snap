// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StrictCspHtmlWebpackPlugin Integration Test should build successfully with Trusted Types enabled 1`] = `
"<!doctype html><html><head><meta http-equiv="Content-Security-Policy" content="script-src 'strict-dynamic' 'sha256-e8lGHsOdJBC9DQQ8Ur9M8w3j6amuteoVxxn7lMC2I/4=' https: 'unsafe-inline';object-src 'none';base-uri 'self';require-trusted-types-for 'script';"><title>Complex Test</title><link rel="preload" href="styles.css" as="style"></head><body><h1>Hello World</h1><script>
    var scripts = [{"src":"library1.bundle.js"},{"src":"app.bundle.js"},{"src":"library2.bundle.js"}];
    var scriptSrcs = new Set(scripts.map(function(s) { return s.src; }));
    var policy = self.trustedTypes && self.trustedTypes.createPolicy ?
      self.trustedTypes.createPolicy('strict-csp#loader', {createScriptURL: function(u) {
        return scriptSrcs.has(u) ? u : null;
      }}) : { createScriptURL: function(u) { return u; } };
    scripts.forEach(function(scriptInfo) {
      var s = document.createElement('script');
      s.src = policy.createScriptURL(scriptInfo.src);
      if (scriptInfo.type) {
        s.type = scriptInfo.type;
      }
      s.async = false; // preserve execution order.
      document.body.appendChild(s);
    });
    </script><script>console.error("No reportUri provided. Trusted Types reports will not be sent to a remote endpoint.")</script></body></html>"
`;

exports[`StrictCspHtmlWebpackPlugin Integration Test should build successfully with Trusted Types in report-only mode 1`] = `
"<!doctype html><html><head><meta http-equiv="Content-Security-Policy" content="script-src 'strict-dynamic' 'sha256-e8lGHsOdJBC9DQQ8Ur9M8w3j6amuteoVxxn7lMC2I/4=' https: 'unsafe-inline';object-src 'none';base-uri 'self';require-trusted-types-for 'script';"><title>Complex Test</title><link rel="preload" href="styles.css" as="style"></head><body><script>
    if (self.trustedTypes && !self.trustedTypes.defaultPolicy) {
      
      const generateAndSendReport = function(sample) {
        const stack = (new Error()).stack;
        const regex = /([^ ()]+):(d+):(d+)/g;
        let match;
        let lastMatch;
        while ((match = regex.exec(stack)) !== null) {
          lastMatch = match;
        }
        const data = JSON.stringify({'csp-report': {
          'document-uri': window.location.href,
          'referrer': '', // No way of knowning this.
          'violated-directive': 'require-trusted-types-for',
          'effective-directive': 'require-trusted-types-for',
          'original-policy': '', // No way of knowing this
          'disposition': 'report',
          'blocked-uri': 'trusted-types-sink',
          'line-number': lastMatch && lastMatch[2] ? +lastMatch[2] : 0,
          'column-number': lastMatch && lastMatch[3] ? +lastMatch[3] : 0,
          'source-file': lastMatch && lastMatch[1] ? lastMatch[1] : '',
          'status-code': 0, // No way of knowing this.
          'script-sample': sample
        }});
        const blob = new Blob([data], {'Content-Type': 'application/json'});
        if (self.navigator && self.navigator.sendBeacon) {
          navigator.sendBeacon('https://example.com/report', blob);
        }
        return data;
      };
      
      self.trustedTypes.createPolicy('default', {
        createHTML: function(s) {
          console.error("[Report Only] Uncaught TypeError: This document requires 'TrustedHTML' assignment: " + s);
          generateAndSendReport('Element innerHTML|' + s);
          return s;
        },
        createScript: function(s) {
          console.error("[Report Only] Uncaught TypeError: This document requires 'TrustedScript' assignment: " + s);
          generateAndSendReport('HTMLScriptElement text|' + s);
          return s;
        },
        createScriptURL: function(s) {
          console.error("[Report Only] Uncaught TypeError: This document requires 'TrustedScriptURL' assignment: " + s);
          generateAndSendReport('HTMLScriptElement src|' + s);
          return s;
        },
      });
    }
    </script><h1>Hello World</h1><script>
    var scripts = [{"src":"library1.bundle.js"},{"src":"app.bundle.js"},{"src":"library2.bundle.js"}];
    var scriptSrcs = new Set(scripts.map(function(s) { return s.src; }));
    var policy = self.trustedTypes && self.trustedTypes.createPolicy ?
      self.trustedTypes.createPolicy('strict-csp#loader', {createScriptURL: function(u) {
        return scriptSrcs.has(u) ? u : null;
      }}) : { createScriptURL: function(u) { return u; } };
    scripts.forEach(function(scriptInfo) {
      var s = document.createElement('script');
      s.src = policy.createScriptURL(scriptInfo.src);
      if (scriptInfo.type) {
        s.type = scriptInfo.type;
      }
      s.async = false; // preserve execution order.
      document.body.appendChild(s);
    });
    </script></body></html>"
`;

exports[`StrictCspHtmlWebpackPlugin Integration Test should build successfully without Trusted Types 1`] = `
"<!doctype html><html><head><meta http-equiv="Content-Security-Policy" content="script-src 'strict-dynamic' 'sha256-fosdR5bZcQQV854t9xSNaZphl31I2w+M/eC0xRqzm3Y=' https: 'unsafe-inline';object-src 'none';base-uri 'self';"><title>Complex Test</title><link rel="preload" href="styles.css" as="style"></head><body><h1>Hello World</h1><script>
    var scripts = [{"src":"library1.bundle.js"},{"src":"app.bundle.js"},{"src":"library2.bundle.js"}];
    scripts.forEach(function(scriptInfo) {
      var s = document.createElement('script');
      s.src = scriptInfo.src;
      if (scriptInfo.type) {
        s.type = scriptInfo.type;
      }
      s.async = false; // preserve execution order.
      document.body.appendChild(s);
    });
    </script></body></html>"
`;
